File: LeetCode/src/linkedlist/ReverseLinkedList.java
package linkedlist;

public class ReverseLinkedList {

    public ListNode reverseList(ListNode head) {

        if (head == null || head.next == null) {
            return head;
        }

        ListNode dummy = null;
        ListNode cur = head;

        while (cur != null) {
            ListNode next = cur.next;
            cur.next = dummy;
            dummy = cur;
            cur = next;
        }

        return dummy;
    }

    public static void main(String[] args) {
        ReverseLinkedList solution = new ReverseLinkedList();

        // Test case
        ListNode head = new ListNode(1);
        head.next = new ListNode(2);
        head.next.next = new ListNode(3);
        head.next.next.next = new ListNode(4);
        head.next.next.next.next = new ListNode(5);

        System.out.print("Original Linked List: ");
        ListNode curr = head;
        while (curr != null) {
            System.out.print(curr.val + " ");
            curr = curr.next;
        }
        System.out.println();

        ListNode reversedHead = solution.reverseList(head);

        System.out.print("Reversed Linked List: ");
        ListNode reversed = reversedHead;
        while (reversed != null) {
            System.out.print(reversed.val + " ");
            reversed = reversed.next;
        }
        System.out.println();
    }
}

File: LeetCode_Python/linkedlist/ReverseLinkedList.py
from typing import Optional
from ListNode import ListNode


class ReverseLinkedList:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head

        pre = None
        while head:
            next_node = head.next
            head.next = pre
            pre = head
            head = next_node

        return pre


if __name__ == '__main__':
    solution = ReverseLinkedList()

    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(3)
    head.next.next.next = ListNode(4)
    head.next.next.next.next = ListNode(5)

    result = solution.reverseList(head)
    print("Expected Output:", "[5, 4, 3, 2, 1]")

    actual = []
    while result:
        actual.append(result.val)
        result = result.next
    print("Actual Output:", actual)
